### ISSUE TITLE
Error using kubeconfig when connecting to a multi-master K3s cluster

### ISSUE THREAD

**Issue by rudimk** [k3s-io/k3s#6525]
**Environmental Info:**
K3s Version: v1.21.14+k3s1

Node(s) CPU architecture, OS, and Version: 
`t3a.medium` instances running Ubuntu 20.04.

Cluster Configuration:
3 server nodes, 1 agent, with an external RDS MySQL database. The nodes are provisioned using AWS Autoscaling Groups.

**Describe the bug:**
This is basically the same as #3226, where it seems a previously-identified race condition in the server nodes prevents each server node from picking up certs from the external DB. The server nodes end up generating their own separate certs, which causes a kubeconfig pointing to a load balancer, to fail with the following error:

`Unable to connect to the server: x509: certificate signed by unknown authority (possibly because of "x509: ECDSA verification failure" while trying to verify candidate authority certificate "k3s-server-ca@1668877051")`

**Steps To Reproduce:**
1. Spin up a K3s cluster with >1 nodes, using an external SQL DB - ensure that the nodes are spun up in parallel or within seconds of each other, and use the `--tls-san` flag to register the LB address in their certs.
2. Download the `kubeconfig` and change the server address to the LB address.
3. Try running `kubectl get nodes` a few times in quick succession.

- Installed K3s: 

**Expected behavior:**
All `kubectl` commands using the `kubeconfig` should work as expected.

**Actual behavior:**
Quite randomly, `kubectl` commands using this kubeconfig break with this error:
`Unable to connect to the server: x509: certificate signed by unknown authority (possibly because of "x509: ECDSA verification failure" while trying to verify candidate authority certificate "k3s-server-ca@1668877051")`

**Additional context / logs:**
Screenshot: https://capture.dropbox.com/a4bSkgbMnvXpSqX4

Additional context: the original issue I filed a year ago was with K3s 1.20.6. I found a comment [here](https://github.com/k3s-io/k3s/issues/3950#issuecomment-967251749) where @briandowns states this fix won't be backported to 1.20 - so I tried multiple times with 1.21. A fix that was suggested a year ago was to introduce a jitter or delay between provisioning the first server node and subsequent server nodes. This unfortunately isn't possible with Terraform and AWS Autoscaling Groups, where the creation of an ASG automatically triggers the entire nodepool to come online in one go, rather than a staggered rollout of nodes.

**Comment by rudimk** [comment]
I was going through some of the backport issues for 1.21, and I found this: https://github.com/k3s-io/k3s/issues/3015#issuecomment-967255745

> The TLDR design of this is: Trust the certs from the DB. If certs on disk do not match, have user remove the unmatching files so that k3s can pull the proper certs from DB.

There's a major problem with this, in scenarios like mine where we're provisioning a K3s cluster using a combination of an autoscaling group and user-data, and we probably won't have SSH access to the underlying nodes - so there's really no way for me to go in and remove the unmatching certs.

**Comment by brandond** [comment]
> This unfortunately isn't possible with Terraform and AWS Autoscaling Groups, where the creation of an ASG automatically triggers the entire nodepool to come online in one go

It is possible. Just add a line like this to your instance userdata script, since cloud-init should run before the k3s service is started:
``bash
sleep $(( ( RANDOM % 10 )  + 1 ))
```

**Comment by rudimk** [comment]
Hey @brandond - thanks for the tip! In the meantime, I figured out another solve - EC Autoscaling allows for lifecycle hooks, which allow me to add an arbitrary delay between each instance addition. Will post an update once I verify it solves our problem here.

**Comment by brandond** [comment]
Yep, there's many different ways to do it, it shouldn't be hard to find one that works for you.

**Comment by rudimk** [comment]
Afraid none of those solutions worked. Lifecycle hooks don't actually wait for individual nodes; they allow the ASG to spin up all the nodes, and then wait. As far as random `sleep` - that didn't work either. I'm now not entirely sure if the problem lies in how K3s is working with certs between the DB and the master nodes, or if it's with how the kubeconfig is being grabbed - for context, my user-data script takes the kubeconfig file at `/etc/rancher/k3s/k3s.yaml`, replaces the `127.0.0.1` portion with the load balancer address and then pushes it to the AWS SSM Parameter Store. But this happens on each node, so technically that single parameter is updated thrice, for three separate kubeconfigs from three separate nodes. 

A couple questions:
1. One would assume that a kubeconfig from _any_ master node should be usable from outside the cluster. Is that not the case?
2. I'm attaching my user-data script here - I wonder if there's something I'm doing wrong here?
```
#!/bin/bash -e

# Allow user supplied pre userdata code

## Install the AWS CLI
DEBIAN_FRONTEND=noninteractive sudo apt update && sudo apt -y install zip unzip python3 python3-venv python-is-python3
curl "https://s3.amazonaws.com/aws-cli/awscli-bundle.zip" -o "awscli-bundle.zip"
unzip awscli-bundle.zip
./awscli-bundle/install -i /usr/local/aws -b /usr/local/bin/aws
rm -rf awscli-bundle
export AWS_REGION=`curl --silent http://169.254.169.254/latest/meta-data/placement/region`
export AWS_DEFAULT_REGION=`curl --silent http://169.254.169.254/latest/meta-data/placement/region`
export AWS_PROVIDER_ID=`curl -s http://169.254.169.254/latest/meta-data/placement/availability-zone)/$(curl -s http://169.254.169.254/latest/meta-data/instance-id)`

## Install the Docker runtime
curl https://releases.rancher.com/install-docker/19.03.sh | sudo sh

## Simulate waits, in order to get around K3s' race condition when multiple server nodes try to connect to the K3s DB simultaneously.
sleep $(( ( RANDOM % 10 )  + 1 ))

## Install and initialise K3s
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v${cluster_version}+k3s1 K3S_TOKEN=${token} K3S_DATASTORE_ENDPOINT="mysql://p3s:${rds_password}@tcp(${rds_hostname}:3306)/k3sdb" INSTALL_K3S_EXEC=`server --tls-san ${lb_endpoint} --tls-san $(curl http://169.254.169.254/latest/meta-data/local-ipv4) --docker --disable-cloud-controller --kubelet-arg="cloud-provider=external" --kubelet-arg="provider-id=aws:///$AWS_PROVIDER_ID" --disable servicelb --disable traefik --disable metrics-server --node-label rudimk.com/workload-kind=system --node-taint rudimk.com/workload-kind=system:NoSchedule` sh -s -


while ! sudo [ -e /etc/rancher/k3s/k3s.yaml ]
do 
  sleep 5
done

sudo cp /etc/rancher/k3s/k3s.yaml /home/ubuntu/kubeconfig.yaml && sudo chown ubuntu:ubuntu /home/ubuntu/kubeconfig.yaml

sed -i "s/127.0.0.1/${lb_endpoint}/g" /home/ubuntu/kubeconfig.yaml

aws ssm put-parameter --name ${cluster_name}-kubeconfig --value "`cat /home/ubuntu/kubeconfig.yaml`" --type String --overwrite

# Allow user supplied userdata code
echo foo bar
```

Really do appreciate the help! :)

**Comment by rudimk** [comment]
Happy to report that I re-ran this again - at first I thought that the `sleep` hack wasn't working. But eventually it was clear I was seeing a duplicate of https://github.com/k3s-io/k3s/issues/4788 - the load balancer address was too big. Fixed that up, and everything's working the way it ought to. Thanks @brandond for your help!
